pkg/personname/info.go

```diff
@@ -0,0 +1,202 @@
+/*
+The personname package provides methods and data types for inspecting Person Name (PN)
+Value Representations, as defined here:
+
+http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html
+*/
+package personname
+
+import (
+	"fmt"
+	"strings"
+)
+
+const groupSep = "="
+
+// pnGroup is an enum value for the PN group types (Alphabetic, Ideographic & Phonetic).
+type pnGroup int
+
+// String representation -- mostly for formatting error messages.
+func (group pnGroup) String() string {
+	switch group {
+	case 0:
+		return "Alphabetic"
+	case 1:
+		return "Ideographic"
+	case 2:
+		return "Phonetic"
+	default:
+		panic(fmt.Errorf("bad pnGroup value: %v", int(group)))
+	}
+}
+
+// Enum values for pnGroup
+const (
+	pnGroupAlphabetic pnGroup = iota
+	pnGroupIdeographic
+	pnGroupPhonetic
+)
+
+// Info holds information from an element with a "PN" VR. See the "PN"
+// entry at: http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html
+//
+// PN values are split into three groups which represent three different ways to
+// represent a name:
+//
+// - Alphabetic: How a name is formally spelled using a Phonetic alphabet.
+// - Ideographic: How a name is represented using ideograms / ideographs.
+// - Phonetic: How a name is pronounced.
+//
+// Each of these groups can be inspected to access their individual segments (family
+// name, Given name, etc.)
+type Info struct {
+	// Alphabetic group information about the Alphabetic group.
+	Alphabetic GroupInfo
+	// Ideographic group information about the Ideographic group.
+	Ideographic GroupInfo
+	// Phonetic group information about the Phonetic group.
+	Phonetic GroupInfo
+
+	// NoNullSeparators will remove repeated separators around null groups when
+	// calling DCM() if set to true.
+	NoNullSeparators bool
+}
+
+// WithNullSeparators returns a new Info object that will keep trailing separators
+// that surround both null groups AND group segments: (ex: 'Potter^Harry^^^==').
+func (info Info) WithNullSeparators() Info {
+	info.NoNullSeparators = false
+	info.Alphabetic.NoNullSeparators = false
+	info.Ideographic.NoNullSeparators = false
+	info.Phonetic.NoNullSeparators = false
+
+	return info
+}
+
+// WithoutNullSeparators returns a new Info object that will remove trailing
+// separators that surround both null groups AND group segments:
+// (ex: 'Potter^Harry').
+func (info Info) WithoutNullSeparators() Info {
+	info.NoNullSeparators = true
+	info.Alphabetic.NoNullSeparators = true
+	info.Ideographic.NoNullSeparators = true
+	info.Phonetic.NoNullSeparators = true
+
+	return info
+}
+
+// dcmRemoveNullStrings removes null group strings from the list of groups to be
+// rendered.
+func (info Info) dcmRemoveNullStrings(groupStrings []string) []string {
```

Comment:
typically I feel that we should declare exported functions before unexported ones for readability, since Go does not care about order of function declaration. (so in this case, we can declare DCM() before this).